ktor-shared/ktor-serialization/ktor-serialization-kotlinx/ktor-serialization-kotlinx-json/jvm/src/KotlinxSerializationJsonJvmConverter.kt

```diff
@@ -0,0 +1,166 @@
+/*
+ * Copyright 2014-2022 JetBrains s.r.o and contributors. Use of this source code is governed by the Apache 2.0 license.
+ */
+
+package io.ktor.serialization.kotlinx.json
+
+import io.ktor.http.*
+import io.ktor.http.content.*
+import io.ktor.serialization.*
+import io.ktor.serialization.kotlinx.*
+import io.ktor.util.*
+import io.ktor.util.reflect.*
+import io.ktor.utils.io.*
+import io.ktor.utils.io.charsets.*
+import io.ktor.utils.io.jvm.javaio.*
+import io.ktor.utils.io.streams.*
+import kotlinx.coroutines.*
+import kotlinx.coroutines.flow.*
+import kotlinx.serialization.*
+import kotlinx.serialization.json.*
+import java.io.*
+import kotlin.reflect.*
+import kotlin.text.*
+
+/**
+ * Creates a JVM converter serializing with the specified string [format] and
+ * [defaultCharset] (optional, usually it is UTF-8).
+ */
+@OptIn(ExperimentalSerializationApi::class, InternalAPI::class)
+public class KotlinxSerializationJsonJvmConverter(
+    private val json: Json
+) : AbstractKotlinxSerializationConverter(json) {
+
+    override suspend fun serializeNullable(
```

Comment:
can it be done through `expect/actual` instead of an abstract class and overriding?

Reply:
@rsinukov I chose to add a `AbstractKotlinxSerializationConverter` only to provide this util function that I need to call in `KotlinxSerializationJsonJvmConverter`
```kotlin
protected fun serializerFromTypeInfo(typeInfo: TypeInfo): KSerializer<*> =
        serializerFromTypeInfo(typeInfo, format.serializersModule)
```
Because I didn't want to have to make `serializerFromTypeInfo(typeInfo: TypeInfo, module: SerializersModule): KSerializer<*>` public : I do not like at all the idea to expose a new public "root" function !

I don't think `expect/actual` would allow me to solve this issue.

I can make `KotlinxSerializationConverter` open so that `KotlinxSerializationJsonJvmConverter` extends it, this would allow me to get rid of `AbstractKotlinxSerializationConverter` and remove the code below and just call "super" for fallback instead
```kotlin
private val fallbackConverter by lazy {
    KotlinxSerializationConverter(json)
}
```
Does this seem ok to you ?

Reply:
Do I understand it right that you extract it to have special handling for JSON? What do you think about keeping all logic internal and check if a format is `Json` inside `KotlinxSerializationConverter` instead of creating a separate class for it? I haven't tried it, but I think it should work.
In the pseudocode, it will be something like this:
```
class KotlinxSerializationConverter {
   override fun serializeNullable(...) {
    if (format !is Json) doBaseSerialization(...)
    else serializeJson(...)
  }
}
```
Btw, serialization can be multiplatform, if instead of `OutputStreamContent` you would use `ChannelWriterContent`. Something like this (simplified):
```
internal fun serializeJson(...) {
  return ChannelWriterContent {
     channel.write("[")
     flow.collect { channel.write(it.serializeToString()); channel.write(","); channel.flush() }
     channel.write("]")
  }
}
``` 

Decoding to a sequence can't be multiplatform, but there you can use `expect/actual` functions
```
override fun deserialize(...) {
    if (format !is Json) doBaseDeserialization(...)
    else deserializeJson(...)
}

internal expect fun deserializeJson(...)

// in jvm 
internal actual fun deserializeJson() { /* do sequence deserialization if needed */ }
// in posix/js
internal actual fun deserializeJson() { /* do base deserialization */ }
```

Right now, with introducing of abstract classes, such a simple thing as converter gets really complex. But what's more important, it breaks binary compatibility, so it can't be merged earlier than 3.0.0, which is not going to be soon.
With the approach of internal check, you also don't need to split `JsonSupport` in between platforms.

Reply:
> Do I understand it right that you extract it to have special handling for JSON?

No, the reason I added an abstract class was simply because I needed to access the method `serializerFromTypeInfo(typeInfo, format.serializersModule)` (currently internal)
 that is in **ktor-serialization-kotlinx** in my JVM specific JSON implementation in the other module **ktor-serialization-kotlinx-json**.

> Right now, with introducing of abstract classes, such a simple thing as converter gets really complex. But what's more important, it breaks binary compatibility, so it can't be merged earlier than 3.0.0, which is not going to be soon.
With the approach of internal check, you also don't need to split JsonSupport in between platforms.

Your suggestion cannot work for this same "between separate modules" problem : `KotlinxSerializationConverter` is in the general **ktor-serialization-kotlinx** module, that does not have a dependency on `kotlinx-serialization-json`, so I cannot make this statement `if (format !is Json)` (Json is not accessible), and cannot access all the optimized Flow / Sequence functions based on InputStream / OutputStream that are also in `kotlinx-serialization-json`.

This not breaking binary compatibility of **ktor-serialization-kotlinx** + my intention to offer Json optimised behavior that is only available from `kotlinx-serialization-json` seems pretty complex :disappointed: 
Maybe I should just offer `Flow<T>` streaming for Gson and Jackson, and stick to the naive Flow support for kotlinx-serialization as below ?
```
// specific behavior for kotlinx.coroutines.flow.Flow : collect it into a List
val resolvedValue = if (typeInfo.type == Flow::class) {
    (value as Flow<*>).toList()
} else {
    value
}
```

Reply:
I could add a `compileOnly` dependency on `kotlinx-serialization-json` in **ktor-serialization-kotlinx** , and replace the `if (format is Json)` check with something like `if (format::class.qualifiedName == "kotlinx.serialization.json.Json")`, and then in this if case I could use `kotlinx-serialization-json`.

Or I can change this method signature, but this would be a binary compatibility change 
```kotlin
public fun Configuration.serialization(
    contentType: ContentType,
    format: StringFormat,
    isJson: Boolean = false // <- add this new parameter
)
```

Reply:
You are absolutely right about modules, and that it's impossible to check if the format is `Json` without `compileOnly` dependency. I missed this, sorry.
Unfortunately, breaking binary compatibility is not an option and we need to find a solution that keeps it. I will experiment with it this week.